pkg/urbit/vere/io/ames.c

```diff
@@ -116,31 +120,195 @@ _ames_pact_free(u3_pact* pac_u)
 /* _ames_panc_free(): remove references, lose refcounts and free struct
 */
 static void
-_ames_panc_free(u3_panc* pac_u) {
+_ames_panc_free(u3_panc* pac_u)
+{
   if ( 0 != pac_u->nex_u ) {
     pac_u->nex_u->pre_u = pac_u->pre_u;
   }
+
   if ( 0 != pac_u->pre_u ) {
     pac_u->pre_u->nex_u = pac_u->nex_u;
-  } else {
+  }
+  else {
     c3_assert(pac_u == pac_u->sam_u->pac_u);
     pac_u->sam_u->pac_u = pac_u->nex_u;
   }
 
-  c3_free(pac_u->bod_u.con_y);
+  c3_free(pac_u->ptr_v);
   c3_free(pac_u);
 }
 
-/* _ca_mug_body(): truncated mug hash of bytes
+/* _ames_mug_body(): truncated (20 least-significant bits) mug hash of bytes
 */
 static c3_l
-_ca_mug_body(c3_w len_w, c3_y* byt_y)
+_ames_mug_body(c3_w len_w, c3_y* byt_y)
 {
-  //  mask off ((1 << 20) - 1)
-  //
   return u3r_mug_bytes(byt_y, len_w) & 0xfffff;
 }
 
+/* _ames_sift_head(): parse packet header.
+*/
+static c3_o
+_ames_sift_head(u3_head* hed_u, c3_y buf_y[4])
+{
+  c3_w hed_w = buf_y[0]
+             | (buf_y[1] <<  8)
+             | (buf_y[2] << 16)
+             | (buf_y[3] << 24);
+
+  //  XX only version 0 currently recognized
+  //
+  hed_u->ver_y = hed_w & 0x7;
+  hed_u->mug_l = (hed_w >>  3) & 0xfffff; // 20 bits
+  hed_u->sac_y = (hed_w >> 23) & 0x3;
+  hed_u->rac_y = (hed_w >> 25) & 0x3;
+  hed_u->enc_o = (hed_w >> 27) & 0x1;
+  return c3y;
+}
+
+/* _ames_etch_head(): serialize packet header.
+*/
+static void
+_ames_etch_head(u3_head* hed_u, c3_y buf_y[4])
+{
+  c3_w hed_w = hed_u->ver_y
+             | (hed_u->mug_l <<  3)
+             | (hed_u->sac_y << 23)
+             | (hed_u->rac_y << 25)
+             | (hed_u->enc_o << 27);
+
+  //  only version 0 currently recognized
+  //
+  c3_assert( 0 == hed_u->ver_y );
+
+  buf_y[0] = hed_w & 0xff;
+  buf_y[1] = (hed_w >>  8) & 0xff;
+  buf_y[2] = (hed_w >> 16) & 0xff;
+  buf_y[3] = (hed_w >> 24) & 0xff;
+}
+
+/* _ames_chub_bytes(): c3_y[8] to c3_d
+** XX move
```

Comment:
To where? Might be in scope for this.

Reply:
There's a bunch of placed were we do this kind byte/integer un/packing -- it'd be nice to factor it out where possible. I've updated the comments.